src/Microsoft.AspNetCore.Sockets.Common/Internal/Formatters/ServerSentEventsMessageParser.cs

```diff
@@ -0,0 +1,193 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.using System;
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace Microsoft.AspNetCore.Sockets.Internal.Formatters
+{
+    public class ServerSentEventsMessageParser
+    {
+        static readonly byte ByteCR = (byte)'\r';
+        static readonly byte ByteLF = (byte)'\n';
+
+        private InternalParseState _internalParserState = InternalParseState.ReadMessageType;
+        private IList<byte[]> _data = new List<byte[]>();
+
+        public ParseResult ParseMessage(ReadableBuffer buffer, out ReadCursor consumed, out ReadCursor examined, out Message message)
+        { 
+            consumed = buffer.Start;
+            examined = buffer.End;
+            message = new Message();
+            var messageType = MessageType.Text;
+            var reader = new ReadableBufferReader(buffer);
+
+            var start = consumed;
+            var end = examined;
+
+            while (!reader.End)
+            {
+                if (ReadCursorOperations.Seek(start, end, out var lineEnd, ByteLF) == -1)
+                {
+                    //Partial message. We need to read more.
+                    return ParseResult.Incomplete;
+                }
+
+                lineEnd = buffer.Move(lineEnd, 1);
+                var line = ConvertBufferToSpan(buffer.Slice(start, lineEnd));
+                reader.Skip(line.Length);
+
+                //Check for a misplaced '\n'
+                if (line.Length <= 1 || line[line.Length - 2] != ByteCR)
+                {
+                    throw new FormatException("There was an issue with the frame format");
+                }
+
+                //Strip the \r\n from the span
+                line = line.Slice(0, line.Length - 2);
+
+                switch (_internalParserState)
+                {
+                    case InternalParseState.ReadMessageType:
+                        messageType = GetMessageType(line);
+                        start = lineEnd;
+                        _internalParserState = InternalParseState.ReadMessagePayload;
+                        consumed = lineEnd;
+
+                        break;
+                    case InternalParseState.ReadMessagePayload:
+                        if (!StartsWithDataPrefix(line))
```

Comment:
You only seem to use  `StartsWithDataPrefix` to throw the exception. Change the method name to EnsureStartsWithDataPrefix and throw directly from this method to avoid throwing different exception for the same condition.